keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,910 @@
+where `|succeededIndexes|` means the number of indexes in the `succeededIndexes`. 
+
+In addition, when the `successPolicy` isn't nil, we validate the `criteria` isn't nil.
+
+### Evaluation
+
+Every time the pod condition are updated, the job-controller evaluates the successPolicies following the rules in order:
+
+- `succeededIndexes`: the job-controller evaluates `.status.completedIndexes` to see if a set of indexes is there.
+- `succeededCount`: the job-controller evaluates `.status.succeeded` to see if the value is `succeededCount` or more.
+
+After that, the job-controller adds a `SuccessCriteriaMet` condition instead of a `Failed` condition to `.status.conditions`
+and the job-controller terminates the lingering pods. At that time, `JobSuccessPolicy` is set to the `status.reason` field.
+
+Note that when the job meets one of successPolicies, other successPolicies are ignored.
+
+Finally, once all pods have terminated, the job-controller adds a `Complete` condition to `.status.conditions`.
+
+If any successPolicy isn't set, the job-controller adds both a `SuccessCriteriaMet` and a `Complete` conditions to the Job
+after the job finished.
+
+### Transition of "status.conditions"
+
+When the job with successPolicies is submitted, the job `status.conditions` transits in the following:
+Note that the Job doesn't have an actual `Running` condition in the `status.conditions`.
+
+```mermaid
+stateDiagram-v2
+    [*] --> Running
+    Running --> Failed: Exceeded backoffLimit\n or Matched FailurePolicy
+    Failed --> [*]
+    Running --> SuccessCriteriaMet: Matched SuccessPolicy
+    SuccessCriteriaMet --> Complete: All pods were terminated by successPolicies
+    Complete --> [*]
+```
+
+It means that the job's `.status.conditions` follows the following rules:  
+
+- The job could have both `SuccessCriteriaMet=true` and `Complete=true` conditions.
+- The job can't have both `Failed=true` and `SuccessCriteriaMet=true` conditions.
+- The job can't have both `Failed=true` and `Complete=true` conditions.
+
+#### The situations where successPolicy conflicts other terminating policies
+
+The successPolicy has potential conflicts with other terminating policies such as the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
+, backoffLimit, and [backoffLimitPerIndex](https://kubernetes.io/docs/concepts/workloads/controllers/job/#backoff-limit-per-index)
+in the following situations:
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
+- when the job meets the successPolicy and the number of failed pods exceeds `backoffLimit`.
+- when the job meets the successPolicy and the number of failed pods per indexes exceeds `backoffLimitPerIndex` in all indexes.
+
+To avoid the above conflicts, terminating policies are evaluated the first before successPolicies.
+This means that the terminating policies are respected rather than the successPolicies.
```

Comment:
```suggestion
This means that the terminating policies are respected rather than the successPolicies,
if the Job doesn't have the `FailureTarget` or `SuccessCriteriaMet` conditions yet.
```

Reply:
Done.